ksql-engine/src/main/java/io/confluent/ksql/function/UdfLoader.java

```diff
@@ -320,6 +327,92 @@ private static Object instantiateUdfClass(final Method method,
     }
   }
 
+  private static Object instantiateUdfClass(final Class udfClass,
+                                            final UdfDescription annotation) {
+    try {
+      return udfClass.newInstance();
+    } catch (final Exception e) {
+      throw new KsqlException("Failed to create instance for UDF="
+              + annotation.name(), e);
+    }
+  }
+
+  private Function<List<Schema>,Schema> handleUdfReturnSchema(
+          final Class theClass,
+          final Method method,
+          final Udf udfAnnotation,
+          final UdfDescription descAnnotation) {
+
+    final String schemaProviderName = method.getAnnotation(Udf.class).schemaProvider();
+
+    if (!schemaProviderName.equals("")) {
+      return handleUdfSchemaProviderAnnotation(schemaProviderName, theClass, method,
+              descAnnotation);
+    }
+
+    final Schema returnSchema = getReturnType(method, udfAnnotation);
+    if (returnSchema.name() != null
+        && returnSchema.name().equals("org.apache.kafka.connect.data.Decimal")
+        && schemaProviderName.equals("")) {
+      throw new KsqlFunctionException("BigDecimal return type is not supported "
+          + "without a schema provider method.");
+    }
+
+    return ignored -> returnSchema;
+
+  }
+
+  private Function<List<Schema>,Schema> handleUdfSchemaProviderAnnotation(
+          final String schemaProviderName,
+          final Class theClass,
+          final Method method,
+          final UdfDescription annotation) {
+
+    Function<List<Schema>, Schema> schemaProviderFunction = null;
+
+    final List<Method> allMethods = new ArrayList<Method>(
+            Arrays.asList(theClass.getDeclaredMethods()));
+
+    // throws exception if cannot find method
+    final Method m = findSchemaProvider(allMethods, schemaProviderName);
+    final Object instance = instantiateUdfClass(theClass, annotation);
+
+    schemaProviderFunction = list -> {
+      return invokeSchemaProviderMethod(instance, m, list, annotation);
+    };
+
+    return schemaProviderFunction;
+  }
+
+  private Method findSchemaProvider(final List<Method> allMethods,
+                                    final String schemaProviderName) {
+    for (final Method m : allMethods) {
+      if (m.isAnnotationPresent(UdfSchemaProvider.class)) {
+        if (m.getName().equals(schemaProviderName)) {
+          return m;
+        }
+      }
+    }
+
+    throw new KsqlFunctionException(String.format(
+            "Cannot find schema provider method with name %s. Please provide a method "
+                    + "declaration when using the annotation schemaProvider for a udf.",
+            schemaProviderName));
+
+  }
+
+  private Schema invokeSchemaProviderMethod(final Object instance, final Method m,
+      final List<Schema> args, final UdfDescription annotation) {
+    try {
+      return (Schema) m.invoke(instance,args);
```

Comment:
```suggestion
      return (Schema) m.invoke(instance, args);
```
nit: having the space makes it easier to diff between `instance.args` and `instance, args`